File: artifacts/libzahl-1.0/translate/src/zunsetup.rs
use crate::*;

pub fn helper_zunsetup_1(i_ref: &mut usize) {
    let mut i = *i_ref;
    unsafe {
        libzahl_set_up = 0;
    }

    // Free all the temporary Z structs' chars fields
    let mut free_chars = |z: &std::sync::Mutex<Z>| {
        let mut z = z.lock().unwrap();
        z.chars = Option::None;
    };

    free_chars(&LIBZAHL_TMP_CMP);
    free_chars(&LIBZAHL_TMP_STR_NUM);
    free_chars(&libzahl_tmp_str_mag);
    free_chars(&libzahl_tmp_str_div);
    free_chars(&LIBZAHL_TMP_STR_REM);
    free_chars(&libzahl_tmp_gcd_u);
    free_chars(&libzahl_tmp_gcd_v);
    free_chars(&libzahl_tmp_sub);
    free_chars(&LIBZAHL_TMP_MODMUL);
    free_chars(&libzahl_tmp_div);
    free_chars(&libzahl_tmp_mod);
    free_chars(&libzahl_tmp_pow_b);
    free_chars(&libzahl_tmp_pow_c);
    free_chars(&libzahl_tmp_pow_d);
    free_chars(&libzahl_tmp_modsqr);
    free_chars(&libzahl_tmp_divmod_a);
    free_chars(&libzahl_tmp_divmod_b);
    free_chars(&libzahl_tmp_divmod_d);
    free_chars(&LIBZAHL_TMP_PTEST_X);
    free_chars(&libzahl_tmp_ptest_a);
    free_chars(&LIBZAHL_TMP_PTEST_D);
    free_chars(&libzahl_tmp_ptest_n1);
    free_chars(&LIBZAHL_TMP_PTEST_N4);

    // Free constant Z structs' chars fields
    free_chars(&LIBZAHL_CONST_1E19);
    free_chars(&LIBZAHL_CONST_1E9);
    free_chars(&libzahl_const_1);
    free_chars(&LIBZAHL_CONST_2);
    free_chars(&libzahl_const_4);

    // Free divmod_ds array elements
    {
        let mut divmod_ds = LIBZAHL_TMP_DIVMOD_DS.lock().unwrap();
        for z in divmod_ds.iter_mut() {
            z.chars = Option::None;
        }
    }

    // Free pool elements
    {
        let mut pool_n = LIBZAHL_POOL_N.lock().unwrap();
        let mut pool = LIBZAHL_POOL.lock().unwrap();
        
        for i in 0..pool.len() {
            while pool_n[i] > 0 {
                pool_n[i] -= 1;
                pool[i] = Option::None;
            }
        }
    }

    *i_ref = i;
}
pub fn zunsetup() {
    if unsafe { libzahl_set_up != 0 } {
        let mut i = 0;
        helper_zunsetup_1(&mut i);
    }
}

File: artifacts/libzahl-1.0/decompose/src/zunsetup.c
// hint:  ['i_ref is a mutable refrence to size_t']
void helper_zunsetup_1(size_t * const i_ref)
{
  size_t i = *i_ref;
  libzahl_set_up = 0;
  free(libzahl_tmp_cmp->chars);
  free(libzahl_tmp_str_num->chars);
  free(libzahl_tmp_str_mag->chars);
  free(libzahl_tmp_str_div->chars);
  free(libzahl_tmp_str_rem->chars);
  free(libzahl_tmp_gcd_u->chars);
  free(libzahl_tmp_gcd_v->chars);
  free(libzahl_tmp_sub->chars);
  free(libzahl_tmp_modmul->chars);
  free(libzahl_tmp_div->chars);
  free(libzahl_tmp_mod->chars);
  free(libzahl_tmp_pow_b->chars);
  free(libzahl_tmp_pow_c->chars);
  free(libzahl_tmp_pow_d->chars);
  free(libzahl_tmp_modsqr->chars);
  free(libzahl_tmp_divmod_a->chars);
  free(libzahl_tmp_divmod_b->chars);
  free(libzahl_tmp_divmod_d->chars);
  free(libzahl_tmp_ptest_x->chars);
  free(libzahl_tmp_ptest_a->chars);
  free(libzahl_tmp_ptest_d->chars);
  free(libzahl_tmp_ptest_n1->chars);
  free(libzahl_tmp_ptest_n4->chars);
  ;
  free(libzahl_const_1e19->chars);
  free(libzahl_const_1e9->chars);
  free(libzahl_const_1->chars);
  free(libzahl_const_2->chars);
  free(libzahl_const_4->chars);
  ;
  for (i = 32; i--;)
    free(libzahl_tmp_divmod_ds[i]->chars);

  for (i = (sizeof(libzahl_pool)) / (sizeof(*libzahl_pool)); i--;)
  {
    while (libzahl_pool_n[i]--)
      free(libzahl_pool[i][libzahl_pool_n[i]]);

    free(libzahl_pool[i]);
  }

  *i_ref = i;
}


/*** DEPENDENCIES:
extern z_t libzahl_tmp_mod
----------------------------
extern z_t libzahl_tmp_ptest_d
----------------------------
extern z_t libzahl_tmp_str_div
----------------------------
extern z_t libzahl_tmp_divmod_d
----------------------------
extern z_t libzahl_tmp_pow_c
----------------------------
extern z_t libzahl_tmp_ptest_a
----------------------------
extern int libzahl_set_up
----------------------------
extern z_t libzahl_tmp_ptest_n4
----------------------------
extern z_t libzahl_tmp_str_num
----------------------------
extern z_t libzahl_tmp_pow_b
----------------------------
extern z_t libzahl_tmp_ptest_x
----------------------------
extern z_t libzahl_const_1
----------------------------
extern z_t libzahl_tmp_sub
----------------------------
extern size_t libzahl_pool_n[(sizeof(size_t)) * 8]
----------------------------
extern z_t libzahl_tmp_modsqr
----------------------------
extern z_t libzahl_tmp_gcd_u
----------------------------
extern z_t libzahl_tmp_ptest_n1
----------------------------
extern z_t libzahl_tmp_str_rem
----------------------------
extern z_t libzahl_tmp_divmod_ds[32]
----------------------------
extern zahl_char_t **libzahl_pool[(sizeof(size_t)) * 8]
----------------------------
extern z_t libzahl_tmp_div
----------------------------
extern z_t libzahl_const_4
----------------------------
extern z_t libzahl_const_1e19
----------------------------
extern z_t libzahl_tmp_cmp
----------------------------
extern z_t libzahl_const_1e9
----------------------------
extern z_t libzahl_tmp_gcd_v
----------------------------
extern z_t libzahl_tmp_pow_d
----------------------------
extern z_t libzahl_tmp_divmod_a
----------------------------
extern z_t libzahl_tmp_str_mag
----------------------------
extern z_t libzahl_tmp_divmod_b
----------------------------
extern z_t libzahl_const_2
----------------------------
extern z_t libzahl_tmp_modmul
----------------------------
***/


void zunsetup(void)
{
  size_t i;
  if (libzahl_set_up)
  {
    helper_zunsetup_1(&i);
  }
}


/*** DEPENDENCIES:
extern int libzahl_set_up
----------------------------
void helper_zunsetup_1(size_t * const i_ref)
{
  size_t i = *i_ref;
  libzahl_set_up = 0;
  free(libzahl_tmp_cmp->chars);
  free(libzahl_tmp_str_num->chars);
  free(libzahl_tmp_str_mag->chars);
  free(libzahl_tmp_str_div->chars);
  free(libzahl_tmp_str_rem->chars);
  free(libzahl_tmp_gcd_u->chars);
  free(libzahl_tmp_gcd_v->chars);
  free(libzahl_tmp_sub->chars);
  free(libzahl_tmp_modmul->chars);
  free(libzahl_tmp_div->chars);
  free(libzahl_tmp_mod->chars);
  free(libzahl_tmp_pow_b->chars);
  free(libzahl_tmp_pow_c->chars);
  free(libzahl_tmp_pow_d->chars);
  free(libzahl_tmp_modsqr->chars);
  free(libzahl_tmp_divmod_a->chars);
  free(libzahl_tmp_divmod_b->chars);
  free(libzahl_tmp_divmod_d->chars);
  free(libzahl_tmp_ptest_x->chars);
  free(libzahl_tmp_ptest_a->chars);
  free(libzahl_tmp_ptest_d->chars);
  free(libzahl_tmp_ptest_n1->chars);
  free(libzahl_tmp_ptest_n4->chars);
  ;
  free(libzahl_const_1e19->chars);
  free(libzahl_const_1e9->chars);
  free(libzahl_const_1->chars);
  free(libzahl_const_2->chars);
  free(libzahl_const_4->chars);
  ;
  for (i = 32; i--;)
    free(libzahl_tmp_divmod_ds[i]->chars);

  for (i = (sizeof(libzahl_pool)) / (sizeof(*libzahl_pool)); i--;)
  {
    while (libzahl_pool_n[i]--)
      free(libzahl_pool[i][libzahl_pool_n[i]]);

    free(libzahl_pool[i]);
  }

  *i_ref = i;
}


----------------------------
***/


